python/zfp.pyx

```diff
@@ -0,0 +1,210 @@
+cimport numpy as np
+
+# export #define's
+HEADER_MAGIC = ZFP_HEADER_MAGIC
+HEADER_META = ZFP_HEADER_META
+HEADER_MODE = ZFP_HEADER_MODE
+HEADER_FULL = ZFP_HEADER_FULL
+
+# export enums
+type_none = zfp_type_none
+type_int32 = zfp_type_int32
+type_int64 = zfp_type_int64
+type_float = zfp_type_float
+type_double = zfp_type_double
+mode_null = zfp_mode_null
+mode_expert = zfp_mode_expert
+mode_fixed_rate = zfp_mode_fixed_rate
+mode_fixed_precision = zfp_mode_fixed_precision
+mode_fixed_accuracy = zfp_mode_fixed_accuracy
+
+
+cpdef dtype_to_ztype(dtype):
+    if dtype == np.int32:
+        return zfp_type_int32
+    elif dtype == np.int64:
+        return zfp_type_int64
+    elif dtype == np.float32:
+        return zfp_type_float
+    elif dtype == np.float64:
+        return zfp_type_double
+    else:
+        raise TypeError("Unknown dtype: {}".format(dtype))
+
+cpdef dtype_to_format(dtype):
+    # format characters detailed here:
+    # https://docs.python.org/2/library/array.html#module-array
+    if dtype == np.int32:
+        return 'i' # signed int
+    elif dtype == np.int64:
+        return 'l' # signed long
+    elif dtype == np.float32:
+        return 'f' # float
+    elif dtype == np.float64:
+        return 'd' # double
+    else:
+        raise TypeError("Unknown dtype: {}".format(dtype))
+
+zfp_to_dtype_map = {
+    zfp_type_int32: np.int32,
+    zfp_type_int64: np.int64,
+    zfp_type_float: np.float32,
+    zfp_type_double: np.float64,
+}
+cpdef ztype_to_dtype(zfp_type ztype):
+    try:
+        return zfp_to_dtype_map[ztype]
+    except KeyError:
+        # TODO: is this the correct type?
+        print "Using np.bytes_"
+        return np.bytes_
+
+cdef size_t sizeof_ztype(zfp_type ztype):
+    if ztype == zfp_type_int32:
+        return sizeof(signed int)
+    elif ztype == zfp_type_int64:
+        return sizeof(signed long long)
+    elif ztype == zfp_type_float:
+        return sizeof(float)
+    elif ztype == zfp_type_double:
+        return sizeof(double)
+    else:
+        return -1
+
+cdef zfp_field* _init_field(np.ndarray arr):
+    shape = arr.shape
+    cdef int ndim = arr.ndim
+    cdef zfp_type ztype = dtype_to_ztype(arr.dtype)
+    cdef zfp_field* field
+    cdef void* pointer = arr.data
+
+    strides = [int(x) / arr.itemsize for x in arr.strides[:ndim]]
+
+    if ndim == 1:
+        field = zfp_field_1d(pointer, ztype, shape[0])
+        zfp_field_set_stride_1d(field, strides[0])
+    elif ndim == 2:
+        field = zfp_field_2d(pointer, ztype, shape[0], shape[1])
+        zfp_field_set_stride_2d(field, strides[0], strides[1])
+    elif ndim == 3:
+        field = zfp_field_3d(pointer, ztype, shape[0], shape[1], shape[2])
+        zfp_field_set_stride_3d(field, strides[0], strides[1], strides[2])
+    elif ndim == 4:
+        field = zfp_field_4d(pointer, ztype, shape[0], shape[1], shape[2], shape[3])
+        zfp_field_set_stride_4d(field, strides[0], strides[1], strides[2], strides[3])
+    else:
```

Comment:
If the numpy array uses C order, which is more likely than not, then we'd expect strides[0] to be largest and strides[3] to be smallest.  zfp compression will traverse memory using strides[0] first, then strides[1], etc., which is likely to result in poor locality of reference.  When the array is C contiguous, we ought to map shape[3] and strides[3] to x, shape[2] and strides[2] to y, etc.

The above works but may not be efficient.

Reply:
Do we want to (by default) do the same thing on the decompression-side too?  That would allow a user to do a "round-trip" through the compressor/decompressor and still have the same data layout at the end of it.

Also, should we try and analyze the numpy strides to determine if they should be flipped or not?  In the off-chance that the numpy array is in Fortran ordering, then the strides should not be reversed for optimal locality, right?

Reply:
To hopefully clarify my previous comment, I am thinking of the case where a user has a "C ordered" numpy array that they want to compress and then decompress.

If they do this round-trip with the current code, the ordering of elements in their decompressed array will match the order in their original array (but the compression will be suboptimal).

If they do this round-trip with the strides in the above code (which only runs on compression) flipped, then the compression will be much better but the ordering of the elements in the decompressed array will not match the ordering of the elements in the original array unless the user explicitly provides the `strides` argument to the decompression function.  This seems like surprising behaviour to me.  I would normally expect a round-trip through compression/decompression to not affect element ordering (or at minimum, for the interface to populate the numpy array with the proper metadata so that it can "lie" to me about the underlying data ordering and provide me with a "view" of the data whose order I expect.)

To get the best of both worlds, maybe we switch to analyzing the strides of the input numpy array so that it is compressed (by default) with the fastest changing dimension first.  Then on decompression, we (by default) decompress assuming that the fastest changing dimension is first.  We can either decompress into "C order" (which seems to be numpy's default/preferred order) or we could decompress in "F order" and set the numpy metadata such that it is aware of this ordering (and then can adjust the view it provides the user depending on what they want).

Hopefully that makes more sense. I may be too close to the problem though and getting myself turned around.  Maybe a face-to-face meeting would help clear things up.

Reply:
I'm confused by your "compression will be suboptimal" comment.  Do you mean that the compression *speed* (not compression ratio) will be suboptimal?  We definitely do need to make sure that numpy and zfp array dimensions agree to get a good compression ratio.  The issue we're discussing here is about the memory layout, not the shape of the array.

Yes, it would be preferable to have the memory layouts agree between compression and decompression, but if the user passes a numpy array to decompress into, then we're at their mercy.  Moreover, nothing is stored in the zfp stream about the source memory layout.

One problem with "analyzing" the strides is that the user may not even be aware of the memory layout of their array.  They know the shape and would expect the same shape returned on decompression.  Note that this shape information may be passed as metadata to some other application, perhaps written in some other language.  If we silently permute the dimensions, then I fear things can go wrong.

I think it's better to just use the strides reported by the numpy array when compressing and to assume (for performance reasons) that the most common case is C order.

On decompression, we may want to allow the user to specify if they want their array in C or Fortran contiguous order.  Of course, the current API allows overriding that with strides.  I think we need to chat in person about how to deal with this.